flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/ChangelogMetricGroup.java

```diff
@@ -0,0 +1,150 @@
+
+    @VisibleForTesting
+    static final String NUMBER_OF_IN_PROGRESS_MATERIALIZATION =
+            PREFIX + ".numberOfInProgressMaterialization";
+
+    @VisibleForTesting
+    static final String NUMBER_OF_COMPLETED_MATERIALIZATION =
+            PREFIX + ".numberOfCompletedMaterialization";
+
+    @VisibleForTesting
+    static final String NUMBER_OF_FAILED_MATERIALIZATION =
+            PREFIX + ".numberOfFailedMaterialization";
+
+    @VisibleForTesting
+    static final String LATEST_FULL_SIZE_OF_MATERIALIZATION =
+            PREFIX + ".lastFullSizeOfMaterialization";
+
+    @VisibleForTesting
+    static final String LATEST_INC_SIZE_OF_MATERIALIZATION =
+            PREFIX + ".lastIncSizeOfMaterialization";
+
+    private final Counter totalMaterializationCounter;
+    private final Counter inProgressMaterializationCounter;
+    private final Counter completedMaterializationCounter;
+    private final Counter failedMaterializationCounter;
+    private final UpdatableGauge<Long> lastFullSizeOfMaterializationGauge;
+    private final UpdatableGauge<Long> lastIncSizeOfMaterializationGauge;
+
+    private final ReentrantLock metricsReadWriteLock = new ReentrantLock();
+
+    ChangelogMetricGroup(MetricGroup parentMetricGroup) {
+        super(parentMetricGroup);
+        this.totalMaterializationCounter = counter(NUMBER_OF_TOTAL_MATERIALIZATION);
+        this.inProgressMaterializationCounter = counter(NUMBER_OF_IN_PROGRESS_MATERIALIZATION);
+        this.completedMaterializationCounter = counter(NUMBER_OF_COMPLETED_MATERIALIZATION);
+        this.failedMaterializationCounter = counter(NUMBER_OF_FAILED_MATERIALIZATION);
+        this.lastFullSizeOfMaterializationGauge =
+                gauge(LATEST_FULL_SIZE_OF_MATERIALIZATION, new SimpleUpdatableGauge<>());
+        this.lastIncSizeOfMaterializationGauge =
+                gauge(LATEST_INC_SIZE_OF_MATERIALIZATION, new SimpleUpdatableGauge<>());
+    }
+
+    void reportPendingMaterialization() {
+        metricsReadWriteLock.lock();
+        try {
+            inProgressMaterializationCounter.inc();
+            totalMaterializationCounter.inc();
+        } finally {
+            metricsReadWriteLock.unlock();
+        }
+    }
+
+    void reportCompletedMaterialization(
+            long fullSizeOfMaterialization, long incSizeOfMaterialization) {
+        metricsReadWriteLock.lock();
+        try {
+            completedMaterializationCounter.inc();
+            if (canDecrementOfInProgressMaterializationNumber()) {
+                inProgressMaterializationCounter.dec();
+            }
+            lastFullSizeOfMaterializationGauge.updateValue(fullSizeOfMaterialization);
+            lastIncSizeOfMaterializationGauge.updateValue(incSizeOfMaterialization);
+        } finally {
+            metricsReadWriteLock.unlock();
+        }
+    }
+
+    void reportFailedMaterialization() {
+        metricsReadWriteLock.lock();
```

Comment:
> The syncrhonization is only needed because the failures are handled by different threads (from the async pool), right?

If we remove in-progress counter, maybe we could not need the syncrhonization again ?

Reply:
The number of failures can still be incremented by different threads IIUC.

Reply:
Thanks!